src/Common/Config/YAMLParser.cpp

```diff
@@ -0,0 +1,183 @@
+#include "YAMLParser.h"
+
+#include <string>
+#include <vector>
+
+#include <Poco/DOM/Document.h>
+#include <Poco/DOM/DOMParser.h>
+#include <Poco/DOM/DOMWriter.h>
+#include <Poco/DOM/NodeList.h>
+#include <Poco/DOM/Element.h>
+#include <Poco/DOM/AutoPtr.h>
+#include <Poco/DOM/NamedNodeMap.h>
+#include <Poco/DOM/Text.h>
+#include <Common/Exception.h>
+
+#include <yaml-cpp/yaml.h>
+
+#include <common/logger_useful.h>
+
+#define YAML_ATTRIBUTE_PREFIX '@'
+
+using namespace Poco::XML;
+
+namespace DB
+{
+
+YAMLParser::YAMLParser() = default;
+
+Poco::AutoPtr<Poco::XML::Document> YAMLParser::parse(const String& path)
+{
+    YAML::Node node_yml;
+    try
+    {
+        node_yml = YAML::LoadFile(path);
+    }
+    catch (const YAML::ParserException&)
+    {
+        /// yaml-cpp cannot parse the file because its contents are incorrect
+        throw Poco::Exception("Unable to parse YAML configuration file ", path);
+    }
+    catch (const YAML::BadFile&)
+    {
+        /// yaml-cpp cannot open the file even though it exists
+        throw Poco::Exception("Unable to open YAML configuration file ", path);
+    }
+    Poco::AutoPtr<Poco::XML::Document> xml = new Document;
+    Poco::AutoPtr<Poco::XML::Element> root_node = xml->createElement("yandex");
+    xml->appendChild(root_node);
+    attrs root_attrs;
+    if (process_node(node_yml, xml, root_node, root_attrs))
+    {
+        throw Poco::Exception("Error while loading config '" + path + "': parser cannot define the type of a node!");
+    }
+    return xml;
+}
+
+void YAMLParser::seq_node_get_attrs(YAML::Node node, attrs& attr_vec)
+{
+    auto it = node.begin();
+    if (it->second.IsScalar())
+    {
+        std::string key, value;
+        key = it->first.as<std::string>();
+        value = it->second.as<std::string>();
+        if (key[0] == YAML_ATTRIBUTE_PREFIX)
+        {
+            attr_vec.push_back(std::make_pair(key.substr(1), value));
```

Comment:
Just `key.substr(1)` looks unclear (because of `1` which looks kind of a magic constant), `key.substr(strlen(YAML_ATTRIBUTE_PREFIX))` is longer but more obvious.

Reply:
Why do we have to check if a key starts with `@`? Are those keys are somehow special?

Reply:
Yes, those keys are converted into XML attributes instead of regular tags ans nodes.